From 54f88a27bbc2c9ace8c32dde38f3d1b8f2bd57bb Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Tue, 01 Feb 2022 16:36:18 +0200
Subject: [PATCH] [SW-69189] gaudi3: disable spi d2d gateway

SPI D2D GW is enabled by default. This patch add support to disable SPI
D2D GW in case of single die SKU.

Change-Id: I8df5c53cb8dcfe1c18eeb33788ca3ac64b6d4539
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 4630a1e..985b56b 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -448,10 +448,13 @@
 	goto *(void *)(ARCBOOT_RAM_HEADER);
 }
 
+__weak void arcboot_rom_disable_spi_gw(void) {}
+
 int arcboot_rom_main(void)
 {
 	arcboot_rom_init();
 	arcboot_rom_fuse();
+	arcboot_rom_disable_spi_gw();
 	arcboot_rom_in_recovery();
 	arcboot_rom_request_cpld_wdog();
 	arcboot_rom_bringup_pcie();
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index ac3f02d..ad6c89a 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -159,3 +159,16 @@
 	if (arcboot_rom_global.config_fuse & EFUSE_SYSCFG_SKU_MASK)
 		arcboot_rom_global.sku = 1;
 }
+
+void arcboot_rom_disable_spi_gw(void)
+{
+	uint32_t val;
+
+	if (!arcboot_rom_global.sku)
+		return;
+
+	/* Disable SPI D2D GW */
+	val = regmap_readl(mmD0_D2D_SPI_GLOBAL_CFG);
+	val &= ~(D0_D2D_SPI_GLOBAL_CFG_MSTR_ENABLE_M);
+	regmap_writel(val, mmD0_D2D_SPI_GLOBAL_CFG);
+}
